src/com/palmergames/bukkit/towny/command/PlotCommand.java

```diff
@@ -1607,33 +1606,39 @@ private boolean handlePlotGroupCommand(String[] split, Player player) throws Tow
 				// Mayor/Assistant of the town.
 				
 				PlotGroup plotGroup = townBlock.getPlotObjectGroup();
-				
 				Runnable permHandler = () -> {
-
-					// setTownBlockPermissions returns a towny permission change object
-					TownyPermissionChange permChange = PlotCommand.setTownBlockPermissions(player, townBlockOwner, townBlock, StringMgmt.remArgs(split, 2));
-					// If the perm change object is not null
-					if (permChange != null) {
-						plotGroup.getPermissions().change(permChange);
-						
-						plotGroup.getTownBlocks().stream()
-							.forEach(tb -> {
-								tb.setPermissions(plotGroup.getPermissions().toString());
-								tb.setChanged(!tb.getPermissions().toString().equals(town.getPermissions().toString()));
-								tb.save();
-								// Change settings event
-								BukkitTools.fireEvent(new TownBlockSettingsChangedEvent(tb));
-							});
-
-						plugin.resetCache();
+					try {
+						// setTownBlockPermissions returns a towny permission change object
+						TownyPermissionChange permChange = PlotCommand.setTownBlockPermissions(player, townBlockOwner, townBlock, StringMgmt.remArgs(split, 2));
+						// If the perm change object is not null
+						if (permChange != null) {
+							plotGroup.getPermissions().change(permChange);
+							plotGroup.getTownBlocks().stream()
+								.forEach(tb -> {
+									try {
+										String permissions = plotGroup.getPermissions().toString();
+										BukkitTools.ifCancelledThenThrow(new TownBlockPermissionChangeEvent(townBlock, permissions));
+										tb.setPermissions(permissions);
+										tb.setChanged(!tb.getPermissions().toString().equals(town.getPermissions().toString()));
+											
+										tb.save();
+										// Change settings event
+										BukkitTools.fireEvent(new TownBlockSettingsChangedEvent(tb));
+									} catch (CancelledEventException ignore) { }

```

Comment:
This one is a bit tricky.

We're throwing one event already on line 1612 (but only for the one plot in a plot group,) and now we're doing the same thing to the plot group. 

The way you've coded this, a cancelled event from line 1612 will stop any further changes.

But a cancelled event on line 1620 will just make the plot group not have a unified permission line, which is something I don't believe we allow. (When plots are added to a plotgroup they take on the groups perm line.)

I think the way to handle this is to loop over the plotgroup and throwing an event on each townblock, before the stream that sets the perm line. That way, if any of the events are cancelled we won't have changed any of the plot group's plot perms.

Reply:
I think it would be better to just exit the method if `permChange` throws an exception. I'll remove the exceptions thrown in the stream. I think this way all plots stay unified should the event be initially cancelled. Let me know if you agree.

Reply:
This will work for your situation, you'd be testing the townblock type, and cancelling the first event on line 1612. Some other setups might be doing things differently though.

Reply:
Should I still throw the exception in the stream then?

Reply:
No you have to do a second stream/loop before the current one. 

That way if there's any exceptions the other stream will never get to alter any of the plots.

Reply:
Alright all done